lib/ddtrace/configuration/agent_settings_resolver.rb

```diff
@@ -0,0 +1,236 @@
+
+      def self.call(settings, logger: Datadog.logger)
+        new(settings, logger: logger).send(:call)
+      end
+
+      private
+
+      attr_reader \
+        :logger,
+        :settings
+
+      def initialize(settings, logger: Datadog.logger)
+        @settings = settings
+        @logger = logger
+      end
+
+      def call
+        AgentSettings.new(
+          ssl: ssl?,
+          hostname: hostname,
+          port: port,
+          timeout_seconds: timeout_seconds,
+          deprecated_for_removal_transport_configuration_proc: deprecated_for_removal_transport_configuration_proc,
+          deprecated_for_removal_transport_configuration_options: deprecated_for_removal_transport_configuration_options
+        )
+      end
+
+      def hostname
+        pick_from(
+          configurations_in_priority_order: [
+            DetectedConfiguration.new(
+              friendly_name: "'settings.tracer.hostname'",
+              value: settings.tracer.hostname
+            ),
+            DetectedConfiguration.new(
+              friendly_name: "#{Datadog::Ext::Transport::HTTP::ENV_DEFAULT_URL} environment variable",
+              value: parsed_url && parsed_url.hostname
+            ),
+            DetectedConfiguration.new(
+              friendly_name: "#{Datadog::Ext::Transport::HTTP::ENV_DEFAULT_HOST} environment variable",
+              value: ENV[Datadog::Ext::Transport::HTTP::ENV_DEFAULT_HOST]
+            )
+          ],
+          or_use_default: Datadog::Ext::Transport::HTTP::DEFAULT_HOST
+        )
+      end
+
+      def port
+        port_from_env = ENV[Datadog::Ext::Transport::HTTP::ENV_DEFAULT_PORT]
+        parsed_port_from_env =
+          if port_from_env
+            begin
+              Integer(port_from_env)
+            rescue ArgumentError
+              log_warning(
+                "Invalid value for #{Datadog::Ext::Transport::HTTP::ENV_DEFAULT_PORT} environment variable " \
+                "('#{port_from_env}'). Ignoring this configuration."
+              )
+            end
+          end
+
+        pick_from(
+          configurations_in_priority_order: [
+            DetectedConfiguration.new(
+              friendly_name: '"settings.tracer.port"',
+              value: settings.tracer.port
+            ),
+            DetectedConfiguration.new(
+              friendly_name: "#{Datadog::Ext::Transport::HTTP::ENV_DEFAULT_URL} environment variable",
+              value: parsed_url && parsed_url.port
+            ),
+            DetectedConfiguration.new(
+              friendly_name: "#{Datadog::Ext::Transport::HTTP::ENV_DEFAULT_PORT} environment variable",
+              value: parsed_port_from_env
+            )
+          ],
+          or_use_default: Datadog::Ext::Transport::HTTP::DEFAULT_PORT
+        )
+      end
+
+      def ssl?
+        !parsed_url.nil? && parsed_url.scheme == 'https'
+      end
+
+      def timeout_seconds
+        Datadog::Ext::Transport::HTTP::DEFAULT_TIMEOUT_SECONDS
```

Comment:
Was the ability to set the transport timeout removed with this change? (Documented here https://github.com/DataDog/dd-trace-rb/blob/master/docs/GettingStarted.md#using-the-nethttp-adapter)

Setting the timeout is currently being used by users in production today.

Reply:
It wasn't removed and will still be honored, although I agree that the reason why is a bit subtle:

Effectively, when we have a `c.tracer.transport_options` proc, because that's an opaque function, it's hard to extract the settings provided therein from the proc, and right now I did not even try.

Thus, the custom timeout, as well as other things you can repeat in the proc (hostname, port, ...) are not visible in the actual `agent_settings` object, and are only carried as part of the `deprecated_for_removal_transport_configuration_proc`.

Because in `lib/ddtrace/transport/http.rb` the `deprecated_for_removal_transport_configuration_proc` is called AFTER everything else, it takes precedence over those values and overwrites them.

This is actually exactly how things worked previously: you could set the hostname via `c.tracer.hostname` or via environment variable (or both) and then supply a `c.tracer.transport_options` proc which would then proceed to overwrite them :)

This is the reason why I call it the `deprecated_for_removal_transport_configuration_proc` -- I think this proc should be put on the chopping block, and a real timeout option (as well as other things you can only configure using the proc) should be added as a setting, which will then be able to be resolved by the `AgentSettingsResolver`.

I've just added a few comments above to try to call this out; let me know if you have any suggestions on how to make this clearer/more obvious.